SOD.Common/Helpers/PlayerStatus.cs

```diff
@@ -0,0 +1,181 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using SOD.Common.Custom;
+using SOD.Common.Extensions;
+using UnityEngine;
+using UniverseLib;
+
+namespace SOD.Common.Helpers
+{
+    public sealed class PlayerStatus
+    {
+        internal PlayerStatus() { }
+
+        /// <summary>
+        /// Raised right before illegal action status will be set (either enabled or disabled).
+        /// </summary>
+        public event EventHandler<IllegalActionStatusArgs> OnBeforeSetIllegalStatus;
+        /// <summary>
+        /// Raised right after illegal action status is successfully set (either enabled or disabled).
+        /// </summary>
+        public event EventHandler<IllegalActionStatusArgs> OnAfterSetIllegalStatus;
+
+        internal void OnSetIllegalActionStatus(IllegalActionStatusArgs eventArgs, bool after)
+        {
+            if (after)
+                OnAfterSetIllegalStatus?.Invoke(this, eventArgs);
+            else
+                OnBeforeSetIllegalStatus?.Invoke(this, eventArgs);
+        }
+
+        internal Dictionary<string, IllegalStatusModifier> IllegalStatusModifierDictionary = new Dictionary<string, IllegalStatusModifier>();
+
+        internal Coroutine UpdateIllegalStatusModifiersCoroutine { get; set; }
+
+        /// <summary>
+        /// Sets an illegal status modifier for the specified duration, or
+        /// indefinitely if the duration is 0.0f or omitted. The modifier will
+        /// cause the Player to have an Illegal Action status while it or any
+        /// other modifier is present. At the end of the duration (if
+        /// provided), the modifier will be automatically removed.
+        /// </summary>
+        /// <param name="key">The key identifying the modifier. Used to overwrite, remove, or toggle the modifier later.</param>
+        /// <param name="force">Whether to overwrite any existing modifier with this key, applying the new duration.</param>
+        /// <param name="duration">The duration that the modifier lasts, in seconds (does not progress while the game is paused).</param>
+        public void SetIllegalStatusModifier(string key, bool force = false, float duration = 0.0f)
```

Comment:
I would prefer duration to be a nullable float and to pass null as default value.
0 gives the idea that it will expire instantly

and force parameter would probably be better to name it overwrite? as the functionality suggests